Helper for writing tests for S3-using code

exarkun opened the pull request:
This is definitely incomplete.  Submitted for feedback on the general approach and comments about what might be missing in terms of tests for this testing helper.

mithrandi commented:
Since master's history was rewritten to fix the email address, this PR had hilarious conflicts; I took the liberty of doing a `git rebase` and force pushing to your branch.

exarkun commented:
Yes, I like the pattern of having a separate object.  I haven't quite ever come across a pattern that makes it easy to implement, though, so I usually don't do it.  Do you have an example I can crib so I don't have to think so hard?

mithrandi commented:
I recently did something like this in `txacme`, if it helps: https://github.com/mithrandi/txacme/blob/bb75f456a541ee0c614d96f861c87f46e44681af/src/txacme/testing.py#L22

mithrandi commented on txaws/testing/s3.py:
```diff
@@ -0,0 +1,102 @@
+        return client, self.get_state(client)
+
+
+class S3ClientState(object):
```
I think this is meant to be named `_S3ClientState`, since that's how it's referred to everywhere else. Given that the tests aren't failing even with this error, it seems like some important test coverage is missing?

mithrandi replied:
I think this is meant to be `_S3ClientState`.

mithrandi commented on txaws/testing/s3.py:
```diff
@@ -0,0 +1,102 @@
+
+
+class _MemoryS3Client(object):
+    _state = ControllerState()
```
This should actually be `_ControllerState`, presumably.

mithrandi replied:
And this should presumably be `_ControllerState`.

mithrandi commented on txaws/testing/service.py:
```diff
@@ -1,12 +1,14 @@
 from txaws.service import AWSServiceEndpoint
 from txaws.testing.ec2 import FakeEC2Client
-
+from txaws.testing.s3 import MemoryS3Client
```
`MemoryS3Client` doesn't exist, `_MemoryS3Client` is presumably the wrong thing, and `MemoryS3` doesn't quite have the expected API here, so I guess this code is unfinished.

mithrandi replied:
This import is still bogus as far as I can tell.

exarkun commented:
Added some tests, fixed the various bugs (many of which you pointed out) they caught.

Also, excitingly, fixed _another_ bug that appears to be present in master and probably makes txAWS S3 completely non-functional??  Against sufficiently new Twisted, anyway, but it doesn't look like you need to have a _particularly_ new version to trigger the problem.  Olden days allowed integer Content-Length header value.  That blows up now.  Live version of the integration test pointed this out.

exarkun commented:
After writing that, I now wonder if a better fix is to _delete_ the `Content-Length`-setting code from txAWS.  I would expect `Agent` to take care of that itself.

mithrandi commented:
From what I recall, the correct way to control Content-Length with Agent is [`IBodyProducer.length`](http://twistedmatrix.com/documents/current/api/twisted.web.iweb.IBodyProducer.html#length) and trying to set the header field directly results in really weird things happening. Glancing at the code again, https://github.com/twisted/twisted/blob/twisted-16.5.0/src/twisted/web/_newclient.py#L674 seems to send the header field unconditionally if `IBodyProducer.length` is present, so you would end up with a request with two Content-Lengths which is definitely nonsense.

mithrandi commented on txaws/testing/s3.py:
```diff
@@ -0,0 +1,138 @@
+
+class S3ClientState(object):
+    """
+    ``_S3ClientState`` instances hold the ``_MemoryS3Client`` instance
```
Extra underscore here in `_S3ClientState`.

mithrandi commented on txaws/s3/tests/test_client.py:
```diff
@@ -1326,3 +1320,44 @@ def test_request_payment_enum(self):
+    # account and dedicating it to testing purposes.
+    try:
+        access_key = environ["TXAWS_INTEGRATION_AWS_ACCESS_KEY_ID"]
+        secret_key = environ["TXAWS_INTEGRATION_AWS_SECRET_ACCESS_KEY"]
```
The secrets aren't available here, because Travis doesn't make secrets avaliable to PR builds; this makes for some unfortunate coverage noise, but I'm not sure what we can do about that.

exarkun replied:
They should be available for twisted/txaws PRs.  I thought it would be good idea to do an exarkun/txaws PR for this.  Clearly not.  In the future we can all just use the twisted/txaws repo, I think.
